src/util/simplify_expr_pointer.cpp

```diff
@@ -474,16 +494,23 @@ bool simplify_exprt::simplify_inequality_pointer_object(exprt &expr)
         return true;
       }
     }
-    else if(op.id() != ID_constant || op.get(ID_value) != ID_NULL)
+    else if(op.id() != ID_constant || !op.is_zero())
     {
       return true;
     }
-  }
 
-  bool equal=expr.op0().op0()==expr.op1().op0();
-
-  expr.make_bool(expr.id()==ID_equal?equal:!equal);
+    if(new_inequality_ops.empty())
+      new_inequality_ops.push_back(op);
+    else
+    {
+      new_inequality_ops.push_back(typecast_exprt::conditional_cast(
+        op, new_inequality_ops.front().type()));
+      simplify_node(new_inequality_ops.back());
```

Comment:
Thought for the future: in this case I think we have a very particular simplification in mind -- presumably that nulls are typed so the cast might be folded, and that `(T*)(S*)x` can become `(T*)x`. However bluntly using the simplifier always risks that a much more expensive transformation might be attempted. This suggests factoring out particular simplifications as utility methods, used both during simplification and as the routine way of creating a pointer cast.

Reply:
That certainly is something to review all the simplifier code for. I actually have some patches in a branch already, but it's not quite a systematic review yet. Will work on this in a separate PR.